pkg/virt-controller/watch/vmi.go

```diff
@@ -1248,10 +1248,26 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
 				}
 			}
 		}
-		if len(vmispec.NetworksToHotplug(vmi.Spec.Networks, vmi.Status.Interfaces)) > 0 {
-			if err := c.handleDynamicInterfaceRequests(vmi.Namespace, vmi.Spec.Domain.Devices.Interfaces, vmi.Spec.Networks, pod); err != nil {
+
+		vmiSpecIfaces := vmispec.FilterInterfacesSpec(vmi.Spec.Domain.Devices.Interfaces, func(iface virtv1.Interface) bool {
+			return iface.State != virtv1.InterfaceStateAbsent
+		})
+		ifacesToHotUnplugExist := len(vmi.Spec.Domain.Devices.Interfaces) > len(vmiSpecIfaces)
+
+		if ifacesToHotUnplugExist && namescheme.PodHasOrdinalInterfaceName(services.NonDefaultMultusNetworksIndexedByIfaceName(pod)) {
+			vmiSpecIfaces = vmi.Spec.Domain.Devices.Interfaces
+			ifacesToHotUnplugExist = false
+			log.Log.Object(vmi).Error("hot-unplug is not supported on old VMIs with ordered pod interface names")
+		}
+		vmiSpecNets := vmi.Spec.Networks
+		if ifacesToHotUnplugExist {
+			vmiSpecNets = vmispec.FilterNetworksByInterfaces(vmi.Spec.Networks, vmiSpecIfaces)
+		}
+		ifacesToHotplugExist := len(vmispec.NetworksToHotplug(vmiSpecNets, vmi.Status.Interfaces)) > 0
```

Comment:
I extracted the logic to a function and placed it in a new file `network.go` with proper testing.
The test is set to the same package as the other tests in this are, to change it much more work is needed to split things apart.

Done

Reply:
Thanks this is much better!

nit: How about renaming the file to `dynamic_network.go`? (so it won't end up a dumpster for any network related to code)

Reply:
when it will be so large that we will feel the pain, then we can split it.
As it stands now, it can accept any networking logic from the existing components, which seems a good first step.